Add helper for area charts

mrosenberg opened the pull request:
References #17 

Creates a helper that can be used in concert with d3-attr 'd' for path elements.

taras commented:
This is awesome. A few small things but others looks great. I'll comment latter today. Quick thought, the component should be in 'ember-sparkles'. What do you think?

mrosenberg commented:
It seems lower level than sparkles in my opinion. We've been running the Addepar charts in our app for the past couple years and it is a great library. We want a more custom approach going forward and its nice to have more direct access to D3.

taras commented:
Why do you think it's lower level?

--
Taras Mankovski
@tarasm




On Mon, Aug 1, 2016 at 9:09 PM +0530, "Matthew Rosenberg" <notifications@github.com> wrote:












It seems lower level than sparkles in my opinion. We've been running the Addepar charts in our app for the past couple years and it is a great library. We want a more custom approach going forward and its nice to have more direct access to D3.



—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.

mrosenberg commented:
It seems the goal of sparkles is to create a chart library. The area helper simply returns the coordinates needed for the d attribute of the path element. I'm really enjoying the d3-helper library as it leaves a lot of the chart structure and design up to my app.

taras commented:
Right, I meant the chart that you made for the demo could be in ember-sparkles. No?

--
Taras Mankovski
@tarasm




On Mon, Aug 1, 2016 at 9:45 PM +0530, "Matthew Rosenberg" <notifications@github.com> wrote:












It seems the goal of sparkles is to create a chart library. The area helper simply returns the coordinates needed for the d attribute of the path element. I'm really enjoying the d3-helper library as it leaves a lot of the chart structure and design up to my app. 



—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.

zigahertz commented:
I think you're both right :)

`d3-area` is quite similar to https://github.com/LocusEnergy/ember-d3-helpers/blob/master/addon/helpers/d3-line.js; while the `area-chart.hbs` demo is something that should live in ember-sparkles

zigahertz commented on tests/dummy/app/helpers/i.js:
```diff
@@ -0,0 +1,9 @@
+
+const matcher = /\$\{\s*(.*?)\s*\}/g;
+
+export function interpolate([ str ], hash) {
```
why did you copy and paste the code from the interpolate helper versus just using the helper in a template?

mrosenberg replied:
I didn't want to add a package dependency w/o asking first. If you want to move the demo to sparkles I can just remove that helper.

zigahertz replied:
yes, I think we should move the demo to ember-sparkles and remove this file (and maybe also the dependency you added in `package.json` ? )

zigahertz replied:
this is all looking great @mrosenberg ! Do you still need this helper and the dependency in package.json? Otherwise - I think this looks good to merge!

mrosenberg commented:
@taras Ah, I misunderstood you. I thought you meant the helper itself. I agree with @zigahertz :)

taras commented on tests/dummy/app/templates/area-chart.hbs:
```diff
@@ -0,0 +1,111 @@
+                    (d3-area
+                      scales.xScale
+                      scales.yScale
+                      xKey=xKey
```
here you can use `(r/get "char")` and `(r/get "count")` once it's in ember-sparkles.

taras commented:
@zigahertz is very agreeable.

taras commented:
I submitted a PR to your PR, to make your PR mergable.

taras commented:
Oh, one thing. We need a README entry for this helper. Add it and we're good to go!

zigahertz commented on addon/helpers/d3-area.js:
```diff
@@ -0,0 +1,12 @@
+export function d3Area( [ xScale, yScale ], { xKey, yKey } ) {
+  let xFn  = (d) => xScale( xKey( d ) );
+  let y1Fn = (d) => yScale( yKey( d ) );
+  let y0Fn = ()  => yScale( 0 );
```
why is `0` hard-coded here?

mrosenberg replied:
Haven't needed to build stacked area charts yet so I left it as zero until a later time.

zigahertz replied:
ah, makes sense, thanks - maybe leave a comment explaining this?

mrosenberg replied:
I'm looking into this as a possible bug. I have charts where the area overlaps the xAxis. I think y0Fn needs to be set to yMin for a single series chart.

zigahertz commented on tests/dummy/app/controllers/area-chart.js:
```diff
@@ -0,0 +1,69 @@
+  }),
+
+
+  xKey( d ) {
```
this could be destructured - 

```
xKey({ char }) {
  return char;
}
```
similar for `yKey`

zigahertz replied:
this file and it's associated HBS file would go in the ember-sparkles demo - also, there, we have started following a convention of `inputKey` and `outputKey` versus `xKey` and `yKey` - check out the source code and a documentation PR I'm currently working on, and please ping me with any questions!

mrosenberg replied:
I'll get these in with @taras changes as soon as I can. Coming up on some deadlines.

zigahertz commented:
@mrosenberg how's this going? do you need some help getting this merged in?

mrosenberg commented:
I'm finishing up a few bug fixes on my project and then I'll have sometime to get this in.

mrosenberg commented:
I merged in the latest from this repo and removed the area route to be added later to Sparkles. I also added a Shape Helper section to the README that documents d3-arc, d3-area, and d3-line. d3-area has an example usage. I'd like to do one for lien and arc once I have time to play with those.

patrickberkeley commented on README.md:
```diff
@@ -31,6 +31,11 @@ Checkout [`ember-sparkles`](https://github.com/LocusEnergy/ember-sparkles) to se
   - [`d3-attr`](#d3-attr)
   - [`d3-call`](#d3-call)
+* [Shape helpers](#selection-helpers)
+  - [`d3-arc`](#d3-arc)
```
Nit, these anchors won't work because the ones generated from the headings will be different.

mrosenberg replied:
Sorry, I don't use markdown often. I'll fix this and the other one.

patrickberkeley commented on README.md:
```diff
@@ -249,6 +254,50 @@ Invokes the specified function exactly once, passing in this selection along wit
+
+
+### `(d3-area [xScale, yScale] {xAccessor, yAccessor})`
+[D3 Arc](https://github.com/d3/d3-shape/blob/master/README.md#areas)
```
Another nit, should this be `D3 Area` ?

patrickberkeley commented on README.md:
````diff
@@ -249,6 +254,50 @@ Invokes the specified function exactly once, passing in this selection along wit
+```
+
+### `(d3-line [xScale, yScale] {xAccessor, yAccessor})`
+[D3 Arc](https://github.com/d3/d3-shape/blob/master/README.md#lines)
````
Same Q on the link text here.

patrickberkeley commented on addon/helpers/d3-area.js:
```diff
@@ -0,0 +1,13 @@
+
+export function d3Area( [ xScale, yScale ], { xAccessor, yAccessor } ) {
+  let yMin = yScale.domain()[0];
+  let xFn  = (d) => xScale(xAccessor(d));
```
Any reason to not include the `x0` and `x1` methods?

mrosenberg replied:
No reason other than time. I'm planning on making a more complex chart in the near future and I will probably work those in then.

patrickberkeley replied:
👍

zigahertz commented:
@mrosenberg looks great! I merged locally due to the README conflicts. Once CI passes I'll publish a new version. fixes #21
